api_app/services/aad_authentication.py

```diff
@@ -317,6 +358,130 @@ def get_workspace_user_emails_by_role_assignment(self, workspace: Workspace):
                     workspace_role_assignments_details[role].append(user.email)
         return workspace_role_assignments_details
 
+    def get_workspace_role_by_name(self, name: str, workspace: Workspace) -> Role:
+        app_roles_endpoint =  f"{MICROSOFT_GRAPH_URL}/v1.0/servicePrincipals/{workspace.properties['sp_id']}/appRoles"
+        graph_data = self._ms_graph_query(app_roles_endpoint, "GET")
+
+        for role in graph_data["value"]:
+            if role["value"] == name:
+                return Role(id=role["id"], value=role["value"],
+                            isEnabled=role["isEnabled"],
+                            description=role["description"],
+                            displayName=role["displayName"],
+                            origin=role["origin"],
+                            allowedMemberTypes=role["allowedMemberTypes"])
+
+        return None
+
+    def get_user_by_email(self, user_email: str) -> User:
+        encoded_email = urllib.parse.quote(user_email)
+        user_by_email_endpoint = f"{MICROSOFT_GRAPH_URL}/v1.0/users/{encoded_email}"
+        user_data = self._ms_graph_query(user_by_email_endpoint, "GET")
+
+        return User(id=user_data["id"][0], name=user_data["displayName"][0], email=user_email, roles=[])
+
+    def assign_workspace_user(self, user: User, workspace: Workspace, role: Role) -> None:
+        # User already has the role, do nothing
+        if self._is_user_in_role(user, role):
+            return
+        if self._is_workspace_role_group_in_use(workspace):
+            return self._assign_workspace_user_to_application_group(user, workspace, role)
+        else:
+            return self._assign_workspace_user_to_application(user, workspace, role)
+
+    def _is_user_in_role(self, user: User, role: Role) -> bool:
+        return any(r for r in user.roles if r == role.value)
+
+    def _is_workspace_role_group_in_use(self, workspace: Workspace) -> bool:
+        msgraph_token = self._get_msgraph_token()
+        application_id = workspace.properties["sp_id"]
+        url = self._get_service_principal_assigned_roles_endpoint(application_id)
+        response = requests.get(url, headers=self._get_auth_header(msgraph_token)).json()
+
+        return any (item for item in response["value"] if item["principalType"] == PrincipalType.Group.value)
+
+    def _assign_workspace_user_to_application_group(self, user: User, workspace: Workspace, role: Role):
+        msgraph_token = self._get_msgraph_token()
+        roles_graph_data = self._get_user_role_assignments(workspace.properties["sp_id"], msgraph_token)
+        sp_graph_data = self._get_app_sp_graph_data(workspace.properties["client_id"])
+        app_id_to_role_name = {app_role["id"]: app_role["value"] for app_role in sp_graph_data["value"][0]["appRoles"]}
+
+        for group in (item for item in roles_graph_data["value"] if item["principalType"] == PrincipalType.Group.value):
```

Comment:
Is this not just going to add the user to a random group that has been assigned to the app role?

Reply:
the terraform uses:

```
resource "azuread_group" "workspace_owners" {
  count            = var.create_aad_groups ? 1 : 0
  display_name     = "${var.workspace_resource_name_suffix} Workspace Owners"
  owners           = [var.workspace_owner_object_id]
  security_enabled = true
}

```

We should maybe output and store the group IDs as workspace properties? If that property exists we add to the group, if it doesn't we add direct to the app role?

Reply:
Interesting, I think when we created this originally we made an assumption that there would only be one group assigned to the app role (for each role type) since that is what the "Add groups" option did.  We can make that more explicit.

Reply:
Yes, we need to, who knows what group might get added to the app role, and then unknowingly other people get added to it and end up with access to something they shouldn't.

Reply:
Do we need to store this in the Cosmos config somewhere or can we just infer the names from convention?

Reply:
I think we need to add the group IDs in outputs.tf for the workspace, and porter.yaml, then the IDs will be stored in cosmos. I think check for existence of that value - will need to be conditional on groups being enabled - and if is there use the ID of the group.

That make sense?

Reply:
Looks like there are already some properties being exported I might be able to use `create_aad_groups` and can then lookup based on the name and the `app_role_id_*` properties.

Reply:
Have added an implementation of this, let me know what you think.